pytorch_lightning/accelerators/accelerator.py

```diff
@@ -35,7 +35,13 @@ def setup_environment(self, root_device: torch.device) -> None:
 
         This is called before the LightningModule/DataModule setup hook which allows the user to access the accelerator
         environment before setup is complete.
+
+        Raises:
+            RuntimeError:
+                If corresponding hardware is not found.
         """
+        if not self.is_available():
```

Comment:
I think this should either be called in the init or not called anywhere and we let each accelerator choose.

We cannot be sure the device won't be accessed before `setup_environment`.

Reply:
I agree, we should first make sure setup_env() is the first method to be called in the run method and omit this check until then. 
Also, I prefer not to have this in init, as instantiating an accelerator should be allowed and running it is a different concern.

Reply:
> We cannot be sure the device won't be accessed before setup_environment. Examples of this are shown in the comments by @rohitgr7

My aim with adding this to `setup_environment` by default is this represents the *latest* time by when the device check should run. This should not preclude earlier possible checks from running

Reply:
@rohitgr7 @awaelchli @carmocca to separate concerns, I can move the check for is_available into a separate PR to unblock this one. Even without this, the rest of the PR is still helpful for the accelerator connector rewrite